src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -302,9 +303,10 @@ public static bool IsUnnecessaryCast(this CastExpressionSyntax cast, SemanticMod
                 cast.Expression, semanticModel, cancellationToken,
                 skipVerificationForReplacedNode: true, failOnOverloadResolutionFailuresInOriginalCode: true);
 
-            // First, check to see if the node ultimately parenting this cast has any
-            // syntax errors. If so, we bail.
-            if (speculationAnalyzer.SemanticRootOfOriginalExpression.ContainsDiagnostics)
+            // First, check to see if the node ultimately parenting this cast has any syntax errors.
+            // If so, we bail. If removing the cast introduces a syntax error, bail out too.
+            if (speculationAnalyzer.SemanticRootOfOriginalExpression.ContainsDiagnostics ||
+                speculationAnalyzer.SemanticRootOfReplacedExpression.ContainsDiagnostics)
```

Comment:
@CyrusNajmabadi what do you think about this approach? are there any concerns with this? or performance concerns?

Reply:
this should probably be done inside speculation analyzer itself in ReplacementChangesSemantics - there are probably other IDE features that might introduce broken code here and doing that would ensure they all behave correctly and don't need to each handle this special case separately

Reply:
another thing i like about this is that it doesn't duplicate the logic of the compiler

Reply:
This seems reasonable to me.

Reply:
@CyrusNajmabadi well it shouldn't 😄 that's all wrong, forget what i said.

the error is not from the parser so this doesn't work... all the tests passed because i didn't pass in c# 7.1 to the tests that expect to not get this offered - it simply wasn't offered because it used c# 7.0

Reply:
turns out manual testing is still useful